### ISSUE TITLE
segmentation fault during video_cycle

### ISSUE THREAD

**Issue by mk6032** [alfredh/baresip#418]
It seems I can reproduce this consistently. Basically I'm using a rpi3 with a usb webcam attached. I place a call to extension 200, which I have registered on a Polycom VVX 500 (this model has color display for video). When the call is established, H.264 seems to be used, the video stream on the VVX 500 is active, yet it is a blank/black stream. I am using the latest 5.7.1 firmware on the Polycom. In the application log of the Polycom, I can see appended as the video begins ..

0528081500|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=1
0528081500|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=2
0528081500|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=4
0528081500|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=8
0528081501|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=16
0528081502|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=32
0528081505|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=64
0528081509|vid  |4|00|[SoVidVdecalDecoderC::doVdecalDecode][ERROR] Failed to decode frame, err='VDECAL_ERR_DEC_STRM'; decodeFailureCount=128

Additionally if I place the call on hold/off hold on the Polycom (sometimes this works out SDP issues), in the application log will be appended ..

0528102244|pps  |4|00|[PpsHybridC]: OnEvCallDelegated - InvalidUserCall
0528102251|vid  |4|00|[SoVidVdecalDecoderC::decode][ERROR] Error! Malformed packet, err='Need SPS NAL unit; Need PPS NAL unit; '; totalMalformedPacket=1; [0x427078d8]
0528102251|vid  |4|00|[SoVidVdecalDecoderC::decode][ERROR] Error! Malformed packet, err='Need SPS NAL unit; Need PPS NAL unit; '; totalMalformedPacket=2; [0x427078d8]
0528102251|vid  |4|00|[SoVidVdecalDecoderC::decode][ERROR] Error! Malformed packet, err='Need SPS NAL unit; Need PPS NAL unit; '; totalMalformedPacket=4; [0x427078d8]
0528102251|vid  |4|00|[SoVidVdecalDecoderC::decode][ERROR] Error! Malformed packet, err='Need SPS NAL unit; Need PPS NAL unit; '; totalMalformedPacket=8; [0x427078d8]

I then use '/video_cycle' while in baresip, which switches to H.263 and corrected the problem. However, I noticed that using /video_cycle will often end in a segfault and core dump. This seems especially true if I use /video_cycle more than once, for example to switch back to H.264.

Hardware is a Raspberry Pi 3 ..

processor	: 3
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 38.40
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4

Hardware	: BCM2835
Revision	: a52082
Serial		: 0000000035a0e841

.. and I am using the debian baresip packages:

$ dpkg -l baresip*
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                        Version            Architecture       Description
+++-===========================-==================-==================-============================================================
ii  baresip                     0.5.0-4            all                portable and modular SIP user-agent - metapackage
ii  baresip-core                0.5.0-4            armhf              portable and modular SIP user-agent - core parts
ii  baresip-ffmpeg              0.5.0-4            armhf              portable and modular SIP user-agent - FFmpeg codecs and form
ii  baresip-gstreamer           0.5.0-4            armhf              portable and modular SIP user-agent - GStreamer pipelines
ii  baresip-gtk                 0.5.0-4            armhf              portable and modular SIP user-agent - GTK+ front-end
ii  baresip-x11                 0.5.0-4            armhf              portable and modular SIP user-agent - X11 features

pi@ntp:~ $ baresip 
baresip v0.5.0 Copyright (C) 2010 - 2016 Alfred E. Heggestad et al.
Local network address:  IPv4=eth0:192.168.2.133 
aucodec: opus/48000/2
aucodec: G722/16000/1
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: vumeter
vidcodec: H264
vidcodec: H263
vidcodec: MP4V-ES
vidcodec: VP8
vidsrc: v4l2
v4l2_codec inited
vidcodec: H264
vidsrc: v4l2_codec
vidsrc: avformat
medianat: stun
medianat: turn
medianat: ice
Populated 1 account
Populated 3 contacts
Populated 4 audio codecs
Populated 1 audio filter
Populated 5 video codecs
Populated 0 video filters
baresip is ready.
205@192.168.1.14: {0/TCP/v4} 200 OK () [1 binding]
All 1 useragent registered successfully! (12 ms)
>                              200
call: connecting to 'sip:200@192.168.1.14;transport=tcp'..
call: SIP Progress: 100 Trying (/)
call: SIP Progress: 100 Trying (/)
call: SIP Progress: 180 Ringing (/)
call: SIP Progress: 180 Ringing (/)
call: SIP Progress: 180 Ringing (/)
call: SIP Progress: 180 Ringing (/)
audio: Set audio decoder: G722 16000Hz 1ch
audio: Set audio encoder: G722 16000Hz 1ch
audio tx pipeline:         src ---> vumeter ---> G722
audio rx pipeline:        play <--- vumeter <--- G722
Set video encoder: H264 packetization-mode=0 (500000 bit/s, 25 fps)
video: no video display
v4l2: /dev/video0: found valid V4L2 device (352 x 288) pixfmt=YUYV
v4l2: Current input: "Camera 1"
Set video decoder: H264 packetization-mode=0
205@192.168.1.14: Call established: sip:200@192.168.1.14;transport=tcp
[libx264 @ 0x6e4256a0] using cpu capabilities: ARMv6 NEON                       [                ]
[libx264 @ 0x6e4256a0] profile High, level 2.0                                  [                ]
Set video encoder: H263  (500000 bit/s, 25 fps)efps=28/0                        [                ]
[libx264 @ 0x6e4256a0] frame I:10    Avg QP:17.79  size: 13535
[libx264 @ 0x6e4256a0] frame P:88    Avg QP:19.40  size:  1195
[libx264 @ 0x6e4256a0] mb I  I16..4: 27.4% 28.2% 44.4%
[libx264 @ 0x6e4256a0] mb P  I16..4:  1.4%  0.8%  0.9%  P16..4: 48.2%  3.8%  2.6%  0.0%  0.0%    skip:42.2%
[libx264 @ 0x6e4256a0] final ratefactor: 16.36
[libx264 @ 0x6e4256a0] 8x8 transform intra:27.8% inter:51.3%
[libx264 @ 0x6e4256a0] coded y,uvDC,uvAC intra: 64.6% 85.6% 77.2% inter: 2.9% 43.7% 13.5%
[libx264 @ 0x6e4256a0] i16 v,h,dc,p: 28% 13%  3% 56%
[libx264 @ 0x6e4256a0] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 17% 11%  5% 12%  8%  9%  7% 16% 15%
[libx264 @ 0x6e4256a0] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 29% 12%  3%  9%  9% 10%  8% 11%  8%
[libx264 @ 0x6e4256a0] i8c dc,h,v,p: 58% 13% 20%  9%
[libx264 @ 0x6e4256a0] Weighted P-Frames: Y:2.3% UV:1.1%
[libx264 @ 0x6e4256a0] kb/s:490.78
Segmentation fault (core dumped)
pi@ntp:~ $ 
[core.zip](https://github.com/alfredh/baresip/files/2041586/core.zip)

I've attached the resulting core file. Please let me know what else I can provide to assist. Thank you for baresip! I've been looking for something like this for a while and it's the best solution I've found.

**Comment by perwx3** [comment]
Suggestion for 'video cycle': I've noticed these segfaults also and i would say that the encoding contexts should be protected with mutex during encoding or else you might get a segfault. Have a look in video.c/encode_rtp_send. Add mutex around the encoding:

lock_write_get(vtx->lock); 
err = vtx->vc->ench(vtx->enc, vtx->picup, frame);
lock_rel(vtx->lock); 

For me, that fixed the problem..

**Comment by perwx3** [comment]
Edit: And this mutex is needed also around mem_deref and encupdh call in video.c/video_encoder_set

lock_write_get(vtx->lock);
vtx->enc = mem_deref(vtx->enc);
err = vc->encupdh(&vtx->enc, vc, &prm, params,
                                  packet_handler, vtx);
lock_rel(vtx->lock);

**Comment by alfredh** [comment]
that is a good observation @perwx3 :)

this branch adds the lock:

https://github.com/alfredh/baresip/tree/video_encoder_lock

**Comment by perwx3** [comment]
Ah, yes, I thought so as well. :)

Good work, Alfred! Your code looks better than mine.

**Comment by alfredh** [comment]
@mk6032 could you please test this PR:

https://github.com/alfredh/baresip/pull/441

**Comment by mk6032** [comment]
Yes certainly. I was hoping to try again after #421 but at the moment still having trouble building/using codecs to cycle through. I will start over with a fresh installation from git. What is the list of packages you need for all audio/video codecs to be available if you're https://github.com/alfredh/baresip/wiki/Install:-GIT-Version ?

**Comment by alfredh** [comment]
you can try this:

```
sudo apt-get install libavcodec-dev libx265-dev libvpx-dev
```

and then rebuild. it should give you the following video codec modules:

```
avcodec.so
h265.so
vp8.so
vp9.so
```

**Comment by mk6032** [comment]
Thanks! I'm no longer able to reproduce the issue.

**Comment by alfredh** [comment]
@mk6032 thanks for reporting and testing :)
